libsolidity/codegen/ir/IRGeneratorForStatements.cpp

```diff
@@ -1361,26 +1394,154 @@ void IRGeneratorForStatements::appendAndOrOperatorCode(BinaryOperation const& _b
+				std::optional<unsigned> offset;
+
+				if (std::holds_alternative<unsigned>(_storage.offset))
+					offset = std::get<unsigned>(_storage.offset);
+
+				IRVariable prepared(type, m_context.newYulVariable());
+				define(prepared, _value);
+
+				m_code <<
+					m_utils.updateStorageValueFunction(type, offset) <<
+					"(" <<
+					_storage.slot <<
+					(
+						std::holds_alternative<string>(_storage.offset) ?
+						(", " + std::get<string>(_storage.offset)) :
+						""
+					) <<
+					", " <<
+					prepared.commaSeparatedList() <<
+					")\n";
+			},
+			[&](IRLValue::Memory const& _memory) {
+				if (type.isValueType())
+				{
+					IRVariable prepared(type, m_context.newYulVariable());
+					define(prepared, _value);
+
+					// TODO: does this actually need cleanup?
+
+					if (_memory.byteArrayElement)
+					{
+						solAssert(type == *TypeProvider::byte(), "");
+						m_code << "mstore8(" + _memory.address + ", byte(0, " + prepared.commaSeparatedList() + "))\n";
+					}
+					else
+						m_code << m_utils.writeToMemoryFunction(type) <<
+							"(" <<
+							_memory.address <<
+							", " <<
+							prepared.commaSeparatedList() <<
+							")\n";
+				}
+				else
+				{
+					solAssert(type.sizeOnStack() == 1, "");
+					solAssert(dynamic_cast<ReferenceType const*>(&type), "");
+					m_code << "mstore(" + _memory.address + ", " + _value.name() + ")\n";
+				}
+			},
+			[&](IRLValue::Stack const& _stack) { assign(_stack.variable, _value); },
+			[&](IRLValue::Tuple const& _tuple) {
+				auto components = std::move(_tuple.components);
+				for (size_t i = 0; i < components.size(); i++)
+				{
+					size_t idx = components.size() - i - 1;
+					if (components[idx])
+					{
+						m_currentLValue.emplace(std::move(*components[idx]));
+						assignCurrentLValue(_value.tupleComponent(idx));
+					}
+				}
+			}
+		},
+		m_currentLValue->kind
+	);
+	m_currentLValue.reset();
+}
+
+IRVariable IRGeneratorForStatements::fetchCurrentLValue()
+{
+	solAssert(!!m_currentLValue, "No current LValue.");
+	auto const& type = m_currentLValue->type;
+	IRVariable result{type, m_context.newYulVariable()};
+	std::visit(GenericVisitor{
+		[&](IRLValue::Storage const& _storage) {
+			if (!type.isValueType())
+				define(result) << _storage.slot << "\n";
+			else if (std::holds_alternative<string>(_storage.offset))
+				define(result) <<
+					m_utils.readFromStorageDynamic(type, false) <<
+					"(" <<
+					_storage.slot <<
+					", " <<
+					std::get<string>(_storage.offset) <<
+					")\n";
+			else
+				define(result) <<
+					m_utils.readFromStorage(type, std::get<unsigned>(_storage.offset), false) <<
+					"(" <<
+					_storage.slot <<
+					")\n";
+		},
+		[&](IRLValue::Memory const& _memory) {
+			if (_memory.byteArrayElement)
+				define(result) <<
+					m_utils.cleanupFunction(type) <<
+					"(mload(" <<
+					_memory.address <<
+					"))\n";
+			else if (type.isValueType())
+				define(result) <<
+					m_utils.readFromMemory(type) <<
```

Comment:
external function type unpacking?

Reply:
Done by ``m_utils.readFromMemory``. That's the beauty of this - as long as the util functions are implemented correctly (which as far as I have seen they seem to be), there's no need for headache about stuff like this - ``result`` knows its type and will consist of two stack slots and the util functions will return two values.